### ISSUE TITLE
How to draw a rawarray Image

### ISSUE THREAD

**Issue by AlexHolly** [godotengine/godot#4024]
Right now I have to save my rawarray(image) first but I want to show it without saving to the drive.

var image_as_raw_array = []#is filled with an png image from www
var path = "res://cache.png"

```
if(f.open( path, File.WRITE )==0):
    f.store_buffer( image_as_raw_array )
    f.close()

    var img = ImageTexture.new()
    img.load( path )

    preview_sprite.set_texture( img )
```

Is there a way without "img.load( path )" something like img.load( rawarray )?

http://www.godotengine.org/qa/744/how-to-draw-a-rawarray-image

**Comment by TheHX** [comment]
For storing a Image in a file you can use ```File::store_var``` and ```File::get_var```.

But if you want create a Image from RawArray now is not pratical. But there's this Image constructor:
```cpp
Image(int p_width, int p_height, int p_mipmaps, Format p_format, const DVector<uint8_t>& p_data);
```
I can bind this constructor GDScript, but i will need increase the maximun number of arguments a constructor can have, it's easy and will not afect anything.

**Comment by AlexHolly** [comment]
Is it possible to read the width and height from the rawarray?
It works with "img.load( path )" so why not with img.load( format, rawarray )

**Comment by puppetmaster-** [comment]
In libgdx I used the [Pixmap class](https://libgdx.badlogicgames.com/nightlies/docs/api/com/badlogic/gdx/graphics/Pixmap.html) to generate and manipulate Image/texture in game. Very fast and very cool to create procedurally generated content.

With new [Texture](https://libgdx.badlogicgames.com/nightlies/docs/api/com/badlogic/gdx/graphics/Texture.html)(pixmap) you got the texture to draw.

Perhaps we could have something similarly in godot.

**Comment by TheHX** [comment]
@AlexHolly, oh i'm sorry i I misunderstood the question, you can transform transform a RawArray in a Image using ```bytes2var```.

@puppetmaster- [Image](http://docs.godotengine.org/en/latest/classes/class_image.html) and [ImageTexture](http://docs.godotengine.org/en/latest/classes/class_image.html). Exposing the additional constructor, maybe, will expand it's possibilities.

**Comment by AlexHolly** [comment]
I think bytes2var only works if the bytes where created by godot. My rawarray is a .png image from www
I try anyway.

**Comment by bojidar-bg** [comment]
bytes2var and var2bytes would (de)serialize a godot Texture, not a `.png` file :smile:

**Comment by Bitiquinho** [comment]
The idea of creating an Image directly from a RawArray looks very useful. At least I can see it working for my case, where I get images from a database. Currently I need to save it to a file locally and load it again as an Image.

**Comment by bojidar-bg** [comment]
Note that this works just fine for C++ (as evidenced by assetlib, even though AL would cache them).

**Comment by cefleet** [comment]
If I could open this up again. 
I was able to expose the constructor to gdscript by by editing the variant_call class
I added the ("data", Variant::RAW_ARRAY):
`_VariantCall::add_constructor(_VariantCall::Image_init1,Variant::IMAGE,"width",Variant::INT,"height",Variant::INT,"mipmaps",Variant::BOOL,"format",Variant::INT,"data",Variant::RAW_ARRAY);`
and adding *p_args[4]:
`r_ret=Image(*p_args[0],*p_args[1],*p_args[2],Image::Format(p_args[3]->operator int()),*p_args[4]);`

and adding then the argument:
`const String& p_name5="", const Variant::Type p_type5=Variant::NIL)`
to the add_constructor function.
So now gdscript allows me to use my RawArray from udp to create the Image.

However it errors because of a size difference. The "data" size that the image expects is much greater than the size of the RawArray. If I save the image and load and 
`img.get_data().size()`
The data.size and the expected size are the same. 
So my question is what do a I need to do to my RawArray before it becomes ready to be image data?
